dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/ops/SharingContext.kt

```diff
@@ -9,21 +9,35 @@ class SharingContext<Initial, Return> {
     internal var areSameType = true
 
     /**
-     * Express the value [toReturn] after the [share] computation is done.
+     * Express the value [toReturn] when the [share] computation is done.
+     * Usually the [toReturn] value is different from the [toSend],
+     * otherwise use [share] without calling this function to return the value [toSend].
      * It can be used with checks after the invocation.
-     *
      * ## Example
      * ```
-     * val res3 = share(0) {f ->
-     *   val minValue = f.toMap().minBy { it.value }.value
-     *   minValue butReturn if (min > 1) "Hello" else null
+     * share(0) {
+     *   sendButReturn(it.toMap().maxBy { v -> v.value }.value, "A string")
      * }
      * ```
      */
-    infix fun Initial.butReturn(toReturn: Return): Return {
-        toBeSent = this
+    fun sendButReturn(toSend: Initial, toReturn: Return): Return {
+        toBeSent = toSend
         toBeReturned = toReturn
         areSameType = false
         return toReturn
     }
+
+    /**
+     * When the [share] computation is done, it evaluates the lambda [toReturn] over [toSend] value and returns its result.
+     * ## Example
+     * ```
+     * share(0) {
+     *   sendButReturn(it.toMap().minBy { v -> v.value }.value) { s ->
+     *      if (s > 1) "Hello" else null
+     *   }
+     * }
+     * ```
+     */
+    fun sendButReturn(toSend: Initial, toReturn: (Initial) -> Return): Return =
+        sendButReturn(toSend, toReturn(toSend))
```

Comment:
```suggestion
    fun sendButReturn(toSend: Initial, toReturn: () -> Return): Return =
        sendButReturn(toSend, toReturn())
```

As @DanySK suggested, since you already have in scope the value to be sent.
Remember to change the tests accordingly.

Reply:
I'm afraid that in this way the compiler has problems with overload.
Applying the suggested changes, I had to update tests as the following:
``` kotlin
share(0) {
    sendButReturn(it.toMap().minBy { v -> v.value }.value) { 
         if (toBeSent!! > 1) "hello" else null 
    }
} // Note that here I had to use `!!` due to toBeSent being a nullable value
```
But it generates the following error:
`Overload resolution ambiguity. All these functions match.
public final fun sendButReturn(toSend: Int, toReturn: () → TypeVariable(Return)): TypeVariable(Return) defined in it.unibo.collektive.aggregate.ops.SharingContext
public final fun sendButReturn(toSend: Int, toReturn: TypeVariable(Return)): TypeVariable(Return) defined in it.unibo.collektive.aggregate.ops.SharingContext`

The original test was:
```kotlin
share(0) {
   sendButReturn(it.toMap().minBy { v -> v.value }.value) { s ->
        if (s > 1) "Hello" else null
   }
}
```
but it has problems inferring the type for parameter `s`.

Reply:
My aim is to replace    
`fun sendButReturn(toSend: Initial, toReturn: (Initial) -> Return): Return`  
with  
`fun sendButReturn(toSend: Initial, toReturn: () -> Return): Return`  
and not having both.

Finally, the two methods will be:
```kotlin
fun sendButReturn(toSend: Initial, toReturn: Return): Return
fun sendButReturn(toSend: Initial, toReturn: () -> Return): Return
```

In this way, no overload conflict must be raised.

So we have two possible usages:

```kotlin
val result = share(0) {
    sendButReturn(10, "hello")
}
```

and

```kotlin
val result = share(0) {
    sendButReturn(10) {
        println("do something")
        "hello"
    }
}
```

There shouldn't be any problems with these two overloads.  
Let me know if there are any.

Reply:
The problem has been solved specifying the return type as following:
```kotlin
val result: String? = share(0) {
    val min = it.toMap().minBy { v -> v.value }.value
    sendButReturn(min) { 
        if (min > 1) "Hello" else null
   }
}
```